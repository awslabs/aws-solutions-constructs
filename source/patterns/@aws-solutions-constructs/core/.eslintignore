index.js
lib/*.js
test/*.js
*.d.ts
coverage
test/integ.*.js.snapshot/
test/cdk-integ.out.integ.*.snapshot
test/lambda/index.js
test/lambda-test/index.js