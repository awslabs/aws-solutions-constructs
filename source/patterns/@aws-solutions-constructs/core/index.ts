/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

export * from './lib/alb-defaults';
export * from './lib/alb-helper';
export * from './lib/apigateway-defaults';
export * from './lib/apigateway-helper';
export * from './lib/dynamodb-table-defaults';
export * from './lib/fargate-defaults';
export * from './lib/fargate-helper';
export * from './lib/iot-topic-rule-defaults';
export * from './lib/kinesis-analytics-defaults';
export * from './lib/kinesis-analytics-helper';
export * from './lib/kinesis-streams-defaults';
export * from './lib/kinesis-streams-helper';
export * from './lib/kinesis-firehose-s3-defaults';
export * from './lib/lambda-event-source-mapping-defaults';
export * from './lib/kms-defaults';
export * from './lib/kms-helper';
export * from './lib/lambda-defaults';
export * from './lib/lambda-helper';
export * from './lib/mediastore-defaults';
export * from './lib/mediastore-helper';
export * from './lib/s3-bucket-defaults';
export * from './lib/s3-bucket-helper';
export * from './lib/secretsmanager-defaults';
export * from './lib/secretsmanager-helper';
export * from './lib/security-group-helper';
export * from './lib/sns-defaults';
export * from './lib/sns-helper';
export * from './lib/sqs-defaults';
export * from './lib/sqs-helper';
export * from './lib/cloudfront-distribution-helper';
export * from './lib/cloudfront-distribution-defaults';
export * from './lib/utils';
export * from './lib/events-rule-defaults';
export * from './lib/cognito-defaults';
export * from './lib/cognito-helper';
export * from './lib/elasticsearch-defaults';
export * from './lib/elasticsearch-helper';
export * from './lib/step-function-defaults';
export * from './lib/step-function-helper';
export * from './lib/dynamodb-table-helper';
export * from './lib/vpc-defaults';
export * from './lib/vpc-helper';
export * from './lib/sagemaker-defaults';
export * from './lib/sagemaker-helper';
export * from './lib/cloudwatch-log-group-helper';
export * from './lib/glue-job-defaults';
export * from './lib/glue-job-helper';
export * from './lib/glue-table-defaults';
export * from './lib/glue-table-helper';
export * from './lib/glue-database-defaults';
export * from './lib/glue-database-helper';
export * from './lib/input-validation';
export * from './test/test-helper';
export * from './lib/ssm-string-parameter-helper';
export * from './lib/eventbridge-helper';
export * from './lib/waf-defaults';
export * from './lib/waf-helper';