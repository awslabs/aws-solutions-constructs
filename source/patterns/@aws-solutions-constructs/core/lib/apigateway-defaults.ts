/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as api from '@aws-cdk/aws-apigateway';
import * as lambda from '@aws-cdk/aws-lambda';
import { LogGroup } from '@aws-cdk/aws-logs';

/**
 * Private function to configure an api.RestApiProps
 * @param scope - the construct to which the RestApi should be attached to.
 * @param _endpointType - endpoint type for Api Gateway e.g. Regional, Global, Private
 * @param _logGroup - CW Log group for Api Gateway access logging
 */
function DefaultRestApiProps(_endpointType: api.EndpointType[], _logGroup: LogGroup): api.RestApiProps {
    return {
        endpointConfiguration: {
            types: _endpointType
        },
        cloudWatchRole: false,
        // Configure API Gateway Access logging
        deployOptions: {
            accessLogDestination: new api.LogGroupLogDestination(_logGroup),
            accessLogFormat: api.AccessLogFormat.jsonWithStandardFields(),
            loggingLevel: api.MethodLoggingLevel.INFO,
            dataTraceEnabled: true,
            tracingEnabled: true
        },
        defaultMethodOptions: {
            authorizationType: api.AuthorizationType.IAM
        }

    } as api.RestApiProps;
}

/**
 * Provides the default set of properties for Edge/Global Lambda backed RestApi
 * @param scope - the construct to which the RestApi should be attached to.
 * @param _endpointType - endpoint type for Api Gateway e.g. Regional, Global, Private
 * @param _logGroup - CW Log group for Api Gateway access logging
 */
export function DefaultGlobalLambdaRestApiProps(_existingLambdaObj: lambda.Function, _logGroup: LogGroup): api.LambdaRestApiProps {
    const baseProps: api.RestApiProps = DefaultRestApiProps([api.EndpointType.EDGE], _logGroup);

    const extraProps: api.LambdaRestApiProps = {
        handler: _existingLambdaObj,
    };

    return Object.assign(baseProps, extraProps);
}

/**
 * Provides the default set of properties for Regional Lambda backed RestApi
 * @param scope - the construct to which the RestApi should be attached to.
 * @param _endpointType - endpoint type for Api Gateway e.g. Regional, Global, Private
 * @param _logGroup - CW Log group for Api Gateway access logging
 */
export function DefaultRegionalLambdaRestApiProps(_existingLambdaObj: lambda.Function, _logGroup: LogGroup): api.LambdaRestApiProps {
    const baseProps: api.RestApiProps = DefaultRestApiProps([api.EndpointType.REGIONAL], _logGroup);

    const extraProps: api.LambdaRestApiProps = {
        handler: _existingLambdaObj,
    };

    return Object.assign(baseProps, extraProps);
}

/**
 * Provides the default set of properties for Edge/Global RestApi
 * @param scope - the construct to which the RestApi should be attached to.
 * @param _endpointType - endpoint type for Api Gateway e.g. Regional, Global, Private
 * @param _logGroup - CW Log group for Api Gateway access logging
 */
export function DefaultGlobalRestApiProps(_logGroup: LogGroup) {
    return DefaultRestApiProps([api.EndpointType.EDGE], _logGroup);
}