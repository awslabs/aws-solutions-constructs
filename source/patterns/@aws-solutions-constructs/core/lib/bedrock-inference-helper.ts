/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from 'aws-cdk-lib';
import * as bedrock from 'aws-cdk-lib/aws-bedrock';
import * as defaults from '../index';
// Note: To ensure CDKv2 compatibility, keep the import statement for Construct separate
import { Construct } from 'constructs';

export interface BuildInferenceProfileProps {
  readonly bedrockModelId: string;
  readonly inferenceProfileProps?: bedrock.CfnApplicationInferenceProfileProps;
  readonly deployCrossRegionProfile?: boolean;
}

export interface BuildInferenceProfileReponse {
  readonly crossRegion?: boolean,
  readonly inferenceProfile: bedrock.CfnApplicationInferenceProfile,
}

export function buildInferenceProfile(scope: Construct, id: string, props: BuildInferenceProfileProps): BuildInferenceProfileReponse {
  const areaMap = createAreaPrefixMapping(scope, id);

  const regionPrefix = cdk.Fn.select(0, cdk.Fn.split('-', cdk.Aws.REGION));
  const crossRegion = IsCrossRegionProfile(props.deployCrossRegionProfile);
  const inferenceSourceArn = crossRegion ?
    `arn:${cdk.Aws.PARTITION}:bedrock:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:inference-profile/${areaMap.mapping.findInMap(regionPrefix, "prefix")}.${props.bedrockModelId}` :
    `arn:${cdk.Aws.PARTITION}:bedrock:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:foundation-model/${props.bedrockModelId}`;

  const constructInferenceProps: bedrock.CfnApplicationInferenceProfileProps = {
    modelSource: {
      copyFrom: inferenceSourceArn
    },
    inferenceProfileName: defaults.generatePhysicalInferenceProfileName("", [id])
  };
  const finalProps = defaults.consolidateProps(constructInferenceProps, props.inferenceProfileProps);
  const inferenceProfile = new bedrock.CfnApplicationInferenceProfile(scope, `${id}-inference`, finalProps);

  return {
    inferenceProfile,
    crossRegion
  };
}

export interface MappingResponse {
  readonly mappingName: string,
  readonly mapping: cdk.CfnMapping,
}

export function createAreaPrefixMapping(scope: Construct, id: string): MappingResponse {
  const mappingName = defaults.removeNonAlphanumeric(`${id}-area-prefix-mapping`);
  const newMapping = new cdk.CfnMapping(scope, `${id}-area-prefix-mapping`, {
    mapping: {
      'us': {
        prefix: 'us',
      },
      'eu': {
        prefix: 'eu',
      },
      'ap': {
        prefix: 'apac',
      },
    }
  });
  newMapping.overrideLogicalId(mappingName);
  return {
    mapping: newMapping,
    mappingName
  };
}

export function createAreaRegionMapping(scope: Construct, id: string, model: string): MappingResponse {
  const mappingName = defaults.removeNonAlphanumeric(`${id}-area-region-mapping`);
  const newMapping = new cdk.CfnMapping(scope, `${id}-area-region-mapping`, {
    mapping: {
      'eu': {
        regionalModels:
          `arn:aws:bedrock:eu-north-1::foundation-model/${model},`+
          `arn:aws:bedrock:eu-central-1::foundation-model/${model},`+
          `arn:aws:bedrock:eu-west-1::foundation-model/${model},`+
          `arn:aws:bedrock:eu-west-3::foundation-model/${model}`
      },
      'us': {
        regionalModels:
          `arn:aws:bedrock:us-east-1::foundation-model/${model},`+
          `arn:aws:bedrock:us-east-2::foundation-model/${model},`+
          `arn:aws:bedrock:us-west-2::foundation-model/${model}`
        },
      'ap': {
        regionalModels:
          `arn:aws:bedrock:ap-southeast-2::foundation-model/${model},`+
          `arn:aws:bedrock:ap-northeast-1::foundation-model/${model},`+
          `arn:aws:bedrock:ap-south-1::foundation-model/${model},`+
          `arn:aws:bedrock:ap-northeast-2::foundation-model/${model},`+
          `arn:aws:bedrock:ap-southeast-1::foundation-model/${model},`+
          `arn:aws:bedrock:ap-northeast-3::foundation-model/${model}`
      }
    }
  });
  newMapping.overrideLogicalId(mappingName);
  return {
    mapping: newMapping,
    mappingName
  };
}

export function IsCrossRegionProfile(deployCrossRegionProfile?: boolean) {
  return defaults.CheckBooleanWithDefault(deployCrossRegionProfile, true);
}

export interface BedrockInferenceProps {
  readonly bedrockModelId: string;
  readonly inferenceProfileProps?: bedrock.CfnApplicationInferenceProfileProps;
  readonly deployCrossRegionProfile?: boolean;
}

export function CheckBedrockInferenceProps(propsObject: BedrockInferenceProps) {
  let errorMessages = '';
  let errorFound = false;

  if (propsObject.inferenceProfileProps?.modelSource) {
    errorMessages += 'Error - The construct will create the modelSource value, it cannot be specified in the props.\n';
    errorFound = true;
  }

  if (errorFound) {
    throw new Error(errorMessages);
  }
}