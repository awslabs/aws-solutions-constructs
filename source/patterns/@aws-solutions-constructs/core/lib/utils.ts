/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as deepmerge from 'deepmerge';
import { flagOverriddenDefaults } from './override-warning-service';
import * as log from 'npmlog';

function isObject(val: object) {
    return val != null && typeof val === 'object'
        && Object.prototype.toString.call(val) === '[object Object]';
}

function isPlainObject(o: object) {
  if (Array.isArray(o) === true) {
    return true;
  }

  if (isObject(o) === false) {
    return false;
  }

  // If has modified constructor
  const ctor = o.constructor;
  if (typeof ctor !== 'function') {
    return false;
  }

  // If has modified prototype
  const prot = ctor.prototype;
  if (isObject(prot) === false) {
    return false;
  }

  // If constructor does not have an Object-specific method
  if (prot.hasOwnProperty('isPrototypeOf') === false) {
    return false;
  }

  // Most likely a plain Object
  return true;
}

function combineMerge(target: any[], source: any[]) {
    return target.concat(source);
}

function overwriteMerge(target: any[], source: any[]) {
  target = source;
  return target;
}

export function overrideProps(DefaultProps: object, userProps: object, concatArray: boolean = false): any {
  // Notify the user via console output if defaults are overridden
  const overrideWarningsEnabled = (process.env.overrideWarningsEnabled !== 'false');
  if (overrideWarningsEnabled) {
    flagOverriddenDefaults(DefaultProps, userProps);
  }
  // Override the sensible defaults with user provided props
  if (concatArray) {
    return deepmerge(DefaultProps, userProps, {
      arrayMerge: combineMerge,
      isMergeableObject: isPlainObject
    });
  } else {
    return deepmerge(DefaultProps, userProps, {
      arrayMerge: overwriteMerge,
      isMergeableObject: isPlainObject
    });
  }
}

export function printWarning(message: string) {
  // Style the log output
  log.prefixStyle.bold = true;
  log.prefixStyle.fg = 'red';
  log.enableColor();
  log.warn('AWS_SOLUTIONS_CONSTRUCTS_WARNING: ', message);
}
