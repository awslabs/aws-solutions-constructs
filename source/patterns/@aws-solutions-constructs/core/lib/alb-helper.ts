/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/*
 *  The functions found here in the core library are for internal use and can be changed
 *  or removed outside of a major release. We recommend against calling them directly from client code.
 */

import * as elb from "aws-cdk-lib/aws-elasticloadbalancingv2";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { ApplicationProtocol, ListenerAction, } from "aws-cdk-lib/aws-elasticloadbalancingv2";
import * as elbt from "aws-cdk-lib/aws-elasticloadbalancingv2-targets";
import { printWarning, consolidateProps } from "./utils";
import { DefaultListenerProps } from "./alb-defaults";
import { createAlbLoggingBucket } from "./s3-bucket-helper";
import { DefaultLoggingBucketProps } from "./s3-bucket-defaults";

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Returns the correct ALB Load Balancer to use in this construct, either an existing
 * one provided as an argument or create  new one otherwise.
 */
export function ObtainAlb(
  scope: Construct,
  id: string,
  vpc: ec2.IVpc,
  publicApi: boolean,
  existingLoadBalancerObj?: elb.ApplicationLoadBalancer,
  loadBalancerProps?: elb.ApplicationLoadBalancerProps | any,
  logAccessLogs?: boolean,
  loggingBucketProps?: s3.BucketProps
): elb.ApplicationLoadBalancer {
  let loadBalancer: elb.ApplicationLoadBalancer;

  if (existingLoadBalancerObj) {
    loadBalancer = existingLoadBalancerObj;
  } else {
    const consolidatedProps = consolidateProps({}, loadBalancerProps, { vpc, internetFacing: publicApi });
    loadBalancer = new elb.ApplicationLoadBalancer(
      scope,
      `${id}-alb`,
      consolidatedProps
    );
    if (logAccessLogs === undefined || logAccessLogs === true) {
      const consolidatedLoggingBucketProps = consolidateProps(DefaultLoggingBucketProps(), loggingBucketProps);
      const loggingBucket = createAlbLoggingBucket(scope, id, consolidatedLoggingBucketProps);
      loadBalancer.logAccessLogs(loggingBucket);
    }
  }
  return loadBalancer;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function AddListener(
  scope: Construct,
  id: string,
  loadBalancer: elb.ApplicationLoadBalancer,
  listenerProps: elb.ApplicationListenerProps | any
): elb.ApplicationListener {
  let consolidatedListenerProps: elb.ApplicationListenerProps;

  consolidatedListenerProps = consolidateProps(DefaultListenerProps(loadBalancer), listenerProps);

  //  create the listener
  const listener = new elb.ApplicationListener(
    scope,
    `${id}-listener`,
    consolidatedListenerProps
  );
  loadBalancer.listeners.push(listener);

  if (consolidatedListenerProps.protocol === elb.ApplicationProtocol.HTTP) {
    // This will use core.printWarning in the actual construct
    printWarning(
      "AWS recommends encrypting traffic to an Application Load Balancer using HTTPS."
    );
    if (listenerProps.certificates?.length > 0) {
      throw new Error("HTTP listeners cannot use a certificate");
    }
  } else {
    if (!listenerProps.certificates || listenerProps.certificates.length === 0) {
      throw new Error("A listener using HTTPS protocol requires a certificate");
    }

    listener.addCertificates(`${id}-cert`, listenerProps.certificates);
  }

  if (consolidatedListenerProps.protocol === elb.ApplicationProtocol.HTTPS) {
    const opt: elb.RedirectOptions = {
      port: "443",
      protocol: "HTTPS",
    };

    const httpListener = new elb.ApplicationListener(
      scope,
      `${id}-redirect`,
      {
        loadBalancer,
        protocol: ApplicationProtocol.HTTP,
        defaultAction: ListenerAction.redirect(opt),
      }
    );
    loadBalancer.listeners.push(httpListener);
  }

  return listener;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Creates a Target Group for Lambda functions and adds the
 * provided functions as a target to that group. Then adds
 * the new Target Group to the provided Listener.
 */
export function AddLambdaTarget(
  scope: Construct,
  id: string,
  currentListener: elb.ApplicationListener,
  lambdaFunction: lambda.IFunction,
  ruleProps?: elb.AddRuleProps,
  targetProps?: elb.ApplicationTargetGroupProps,
): elb.ApplicationTargetGroup  {

  //  Create the target and assign it to a new target group
  const lambdaTarget = new elbt.LambdaTarget(lambdaFunction);
  const newTargetGroup = new elb.ApplicationTargetGroup(scope, `${id}-tg`, {
    targets: [lambdaTarget],
    targetGroupName: targetProps ? targetProps.targetGroupName : undefined,
    healthCheck: targetProps ? targetProps.healthCheck : undefined
  });

  // The interface AddRuleProps includes conditions and priority, combine that
  // with targetGroups and we can assemble AddApplicationTargetGroupProps
  const consolidatedTargetProps = consolidateProps({}, ruleProps, { targetGroups: [newTargetGroup] });
  currentListener.addTargetGroups(`${scope.node.id}-targets`, consolidatedTargetProps);

  newTargetGroup.setAttribute('stickiness.enabled', undefined);
  return newTargetGroup;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function AddFargateTarget(
  scope: Construct,
  id: string,
  currentListener: elb.ApplicationListener,
  fargateService: ecs.FargateService,
  ruleProps?: elb.AddRuleProps,
  targetProps?: elb.ApplicationTargetGroupProps,
): elb.ApplicationTargetGroup  {

  if (targetProps?.protocol !== elb.ApplicationProtocol.HTTPS) {
    printWarning('AWS recommends using HTTPS protocol for Target Groups in production applications');
  }

  const newTargetGroup = new elb.ApplicationTargetGroup(scope, `${id}-tg`, targetProps);

  // The interface AddRuleProps includes conditions and priority, combine that
  // with targetGroups and we can assemble an AddApplicationTargetGroupProps object
  const consolidatedTargetProps = consolidateProps({ targetGroups: [newTargetGroup] }, ruleProps);

  currentListener.addTargetGroups(`${scope.node.id}-targets`, consolidatedTargetProps);
  newTargetGroup.addTarget(fargateService);

  return newTargetGroup;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Looks for the listener associated with Target Groups
 * If there is a single listener, this returns it whether it is HTTP or HTTPS
 * If there are 2 listeners, it finds the HTTPS listener (we assume the HTTP listener redirects to HTTPS)
 */
export function GetActiveListener(listeners: elb.ApplicationListener[]): elb.ApplicationListener {
  let listener: elb.ApplicationListener;

  if (listeners.length === 0 ) {
    throw new Error(`There are no listeners in the ALB`);
  }
  if (listeners.length === 1 ) {
    listener = listeners[0];
  } else {
    listener = listeners.find(i => (i.node.children[0] as elb.CfnListener).protocol === "HTTPS") as elb.ApplicationListener;
  }
  return listener;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function CheckAlbProps(props: any) {
  let errorMessages = '';
  let errorFound = false;

  if (props.listenerProps?.certificateArns) {
    errorMessages += "certificateArns is deprecated. Please supply certificates using props.listenerProps.certificates\n";
    errorFound = true;
  }

  if (
    ((props.existingLoadBalancerObj &&
      props.existingLoadBalancerObj.listeners.length === 0) ||
      !props.existingLoadBalancerObj) &&
    !props.listenerProps
  ) {
    errorMessages += "When adding the first listener and target to a load balancer, listenerProps must be specified and include at least a certificate or protocol: HTTP\n";
    errorFound = true;
  }

  if (
    props.existingLoadBalancerObj &&
    props.existingLoadBalancerObj.listeners.length > 0 &&
    props.listenerProps
  ) {
    errorFound = true;
    errorMessages += "This load balancer already has a listener, listenerProps may not be specified\n";
  }

  if (
    props.existingLoadBalancerObj &&
    props.existingLoadBalancerObj.listeners.length > 0 &&
    !props.ruleProps
  ) {
    errorFound = true;
    errorMessages += "When adding a second target to an existing listener, there must be rules provided\n";
  }

  // Check construct specific invalid inputs
  if (props.existingLoadBalancerObj && !props.existingVpc) {
    errorFound = true;
    errorMessages += "An existing ALB is already in a VPC, that VPC must be provided in props.existingVpc for the rest of the construct to use.\n";
  }

  if (props.loadBalancerProps?.vpc) {
    errorFound = true;
    errorMessages += 'Specify any existing VPC at the construct level, not within loadBalancerProps.\n';
  }

  if (props.existingLoadBalancerObj) {
    printWarning(
      "The public/private property of an existing ALB must match the props.publicApi setting provided."
    );
  }

  if (errorFound) {
    throw new Error(errorMessages);
  }
}