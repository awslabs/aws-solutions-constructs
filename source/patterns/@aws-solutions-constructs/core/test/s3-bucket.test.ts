/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { SynthUtils } from '@aws-cdk/assert';
import { Duration, Stack } from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as kms from '@aws-cdk/aws-kms';
import * as defaults from '../index';
import { overrideProps } from '../lib/utils';
import '@aws-cdk/assert/jest';
import { StorageClass } from '@aws-cdk/aws-s3/lib/rule';

test('s3 bucket with default params', () => {
  const stack = new Stack();

  /** Default Life Cycle policy to transition older versions to Glacier after 90 days */
  const lifecycleRules: s3.LifecycleRule[] = [{
    noncurrentVersionTransitions: [{
        storageClass: StorageClass.GLACIER,
        transitionAfter: Duration.days(90)
    }]
  }];

  new s3.Bucket(stack, 'test-s3-defaults', defaults.DefaultS3Props(undefined, lifecycleRules));
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('test s3Bucket override versioningConfiguration', () => {
  const stack = new Stack();
  const defaultProps: s3.BucketProps = defaults.DefaultS3Props();

  const inProps: s3.BucketProps = {
    versioned: false
  };

  const outProps = overrideProps(defaultProps, inProps);
  new s3.Bucket(stack, 'test-s3-verioning', outProps);

  expect(stack).toHaveResource("AWS::S3::Bucket", {
    BucketEncryption: {
      ServerSideEncryptionConfiguration: [
        {
          ServerSideEncryptionByDefault: {
            SSEAlgorithm: "AES256"
          }
        }
      ]
    },
    PublicAccessBlockConfiguration: {
      BlockPublicAcls: true,
      BlockPublicPolicy: true,
      IgnorePublicAcls: true,
      RestrictPublicBuckets: true
    }
  });
});

test('test s3Bucket override bucketEncryption', () => {
  const stack = new Stack();
  const defaultProps: s3.BucketProps = defaults.DefaultS3Props();

  const inProps: s3.BucketProps = {
    encryption: s3.BucketEncryption.KMS,
    encryptionKey: new kms.Key(stack, 'mykeyid')
  };

  const outProps = overrideProps(defaultProps, inProps);
  new s3.Bucket(stack, 'test-s3-encryption', outProps);

  expect(stack).toHaveResource("AWS::S3::Bucket", {
    BucketEncryption: {
      ServerSideEncryptionConfiguration: [
        {
          ServerSideEncryptionByDefault: {
            KMSMasterKeyID: {
              "Fn::GetAtt": [
                "mykeyidFA4203B0",
                "Arn"
              ]
            },
            SSEAlgorithm: "aws:kms"
          }
        }
      ]
    },
  });
});

test('test s3Bucket override publicAccessBlockConfiguration', () => {
  const stack = new Stack();
  const defaultProps: s3.BucketProps = defaults.DefaultS3Props();

  const inProps: s3.BucketProps = {
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ACLS
  };

  const outProps = overrideProps(defaultProps, inProps);
  new s3.Bucket(stack, 'test-s3-publicAccessBlock', outProps);

  expect(stack).toHaveResource("AWS::S3::Bucket", {
    PublicAccessBlockConfiguration: {
      BlockPublicAcls: true,
      IgnorePublicAcls: true
    },
  });
});

test('test s3Bucket add lifecycleConfiguration', () => {
  const stack = new Stack();
  const defaultProps: s3.BucketProps = defaults.DefaultS3Props();

  const inProps: s3.BucketProps = {
    lifecycleRules: [{
      expiration: Duration.days(365)
    }]
  };

  const outProps = overrideProps(defaultProps, inProps);
  new s3.Bucket(stack, 'test-s3-lifecycle', outProps);

  expect(stack).toHaveResource("AWS::S3::Bucket", {
    LifecycleConfiguration: {
      Rules: [
        {
          Status: 'Enabled',
          ExpirationInDays: 365,
        }
      ]
    }
  });
});

test('test s3Bucket override serverAccessLogsBucket', () => {
  const stack = new Stack();

  const myLoggingBucket: s3.Bucket = new s3.Bucket(stack, 'MyS3LoggingBucket', defaults.DefaultS3Props());

  const myS3Props: s3.BucketProps = defaults.DefaultS3Props(myLoggingBucket);

  defaults.buildS3Bucket(stack, {
    bucketProps: myS3Props
  });

  expect(stack).toHaveResource("AWS::S3::Bucket", {
    LoggingConfiguration: {
      DestinationBucketName: {
        Ref: "MyS3LoggingBucket119BE896"
      }
    }
  });
});