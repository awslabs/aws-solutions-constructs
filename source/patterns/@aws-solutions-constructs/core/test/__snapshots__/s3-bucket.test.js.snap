// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`s3 bucket with default params 1`] = `
Object {
  "Resources": Object {
    "tests3defaults80430774": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
